controllers/hardwareclassification_controller.go

```diff
@@ -17,36 +14,222 @@ package controllers
+				(extractedProfile.NIC == (hwcc.NIC{})) && (extractedProfile.RAM == (hwcc.RAM{})) {
+				err = errors.New("atleast one of the configuration should be provided")
+				break
+			}
+
+			if extractedProfile.CPU != (hwcc.CPU{}) {
+				if extractedProfile.CPU.MinimumCount < 0 || extractedProfile.CPU.MaximumCount < 0 {
+					err = errors.New("enter valid CPU Count")
+					break
+				}
+
+				if extractedProfile.CPU.MinimumSpeed != "" {
+					if minSpeed, error := strconv.ParseFloat(extractedProfile.CPU.MinimumSpeed, 8); error != nil || minSpeed < 0 {
+						err = errors.New("enter valid Minimum ClockSpeed")
+						break
+					}
+				}
+				if extractedProfile.CPU.MaximumSpeed != "" {
+					if _, error := strconv.ParseFloat(extractedProfile.CPU.MaximumSpeed, 8); error != nil {
+						err = errors.New("enter valid Maximum ClockSpeed")
+						break
+					}
+				}
+
+				introspectionDetails["CPU"] = host.Status.HardwareDetails.CPU
+			}
+
+			if extractedProfile.Disk != (hwcc.Disk{}) {
+				if extractedProfile.Disk.MinimumCount < 0 || extractedProfile.Disk.MaximumCount < 0 {
+					err = errors.New("enter valid Disk Count")
+					break
+				}
+
+				if extractedProfile.Disk.MinimumIndividualSizeGB < 0 || extractedProfile.Disk.MaximumIndividualSizeGB < 0 {
+					err = errors.New("enter valid Disk Size in GB")
+					break
+				}
+				introspectionDetails["Disk"] = host.Status.HardwareDetails.Storage
+			}
+
+			if extractedProfile.NIC != (hwcc.NIC{}) {
+				if extractedProfile.NIC.MinimumCount > 0 || extractedProfile.NIC.MaximumCount > 0 {
+					introspectionDetails["NIC"] = host.Status.HardwareDetails.NIC
+				} else {
+					err = errors.New("enter valid NICS Count")
+					break
+				}
+			}
+
+			if extractedProfile.RAM != (hwcc.RAM{}) {
+				if extractedProfile.RAM.MinimumSizeGB > 0 || extractedProfile.RAM.MaximumSizeGB > 0 {
+					introspectionDetails["RAMMebibytes"] = host.Status.HardwareDetails.RAMMebibytes
+				} else {
+					err = errors.New("enter valid RAM size in GB")
+					break
+				}
+			}
+
+			if len(introspectionDetails) > 0 {
+				extractedHardwareDetails[host.ObjectMeta.Name] = introspectionDetails
+			}
+		}
+
+	}
+	if err != nil {
+		return extractedHardwareDetails, err
+	}
+
+	return extractedHardwareDetails, nil
+}
+
+// SetupWithManager will add watches for this controller
 func (r *HardwareClassificationReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&metal3iov1alpha1.HardwareClassification{}).
+		For(&hwcc.HardwareClassification{}).
+		Watches(
+			&source.Kind{Type: &bmh.BareMetalHost{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: handler.ToRequestsFunc(r.BareMetalHostToHardwareClassification),
+			},
+		).
 		Complete(r)
 }
+
+// BareMetalHostToHardwareClassification will return a reconcile request for a
+// HardwareClassification if the event is for a BareMetalHost.
```

Comment:
This seems backwards. We want to watch for HardwareClassification changes and map those to all hosts. There is no need to filter the list of hosts since Reconcile() can exit early if the host has a bad state at the point that it is invoked.

Reply:
Implemeneted watcher on HardwareClassification kind to watch changes and map those to all hosts.
Addressed in new commit: c239dc6